File: apps/web/test/schema-contract.test.ts
// schema-contract.test.ts is the web-side half of the cross-language
// conformance tripwire (design spec §3, "Codegen fidelity"). Before this
// file existed, packages/schema/test/conformance.test.ts already proved
// the generated TS union (packages/schema/gen/ts/resume.ts) matches
// resume.schema.json — but that suite runs only inside packages/schema.
// Nothing forced apps/web itself to depend on @aboutme/schema at all: the
// web app could grow its own parallel section-type model, drift from the
// schema, and packages/schema's conformance suite would stay green the
// whole time, because it was never testing the web app.
//
// This file closes that gap from the web app's side. It has two halves,
// matching two different, independent failure modes, both exercised by
// `npm run test`:
//
//   1. Compile-time: exhaustivenessFixtureSource() builds a throwaway .ts
//      fixture — a switch with a case per KNOWN_SECTION_TYPES entry and a
//      `default` branch assigning the unmatched value to a `never`-typed
//      binding — that imports the REAL generated `Section` type from
//      '@aboutme/schema' (a genuine dependency — see package.json's
//      "file:../../packages/schema" entry, resolved through normal
//      `npm install`/`npm ci`, not a copy-pasted type). The
//      "compiles the real Section union" test below writes it to disk and
//      runs `tsc --noEmit --strict` on it directly (mirroring
//      packages/schema/test/conformance.test.ts's
//      assertTsRejectsForeignField pattern), deliberately bypassing the
//      project tsconfig with `--ignoreConfig` rather than going through
//      `npm run typecheck`: Nuxt's generated tsconfig.app.json only
//      `include`s `test/nuxt/**/*`, not this directory, so vue-tsc never
//      sees apps/web/test/*.test.ts at all (confirmed empirically —
//      `npm run typecheck` alone does NOT catch a missing case here). If
//      packages/schema/gen/ts/resume.ts ever grows a ninth `Section`
//      variant not in KNOWN_SECTION_TYPES, the fixture's switch no longer
//      exhausts `Section`, the `never` assignment fails to compile, tsc
//      exits non-zero, and this vitest test fails.
//   2. Runtime: sectionTypesFromSchemaFile() reads resume.schema.json off
//      disk directly (independently of both generate.mjs and
//      packages/schema/test/conformance.test.ts's own derivation — the
//      point is to check the pipeline against the schema file itself,
//      not trust a derivation already trusted elsewhere) and compares it
//      against KNOWN_SECTION_TYPES, the same literal set both dispatch()
//      and the compile-time fixture above are built from. A sectionType
//      added to the schema shows up here even before anyone updates this
//      file, so `npm run test` fails loudly instead of silently ignoring
//      the new type at runtime.
import { execFileSync } from 'node:child_process';
import { readFileSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import type { Section } from '@aboutme/schema';

const here = dirname(fileURLToPath(import.meta.url));
// apps/web/test -> apps/web -> apps -> repo root -> packages/schema/...
const schemaFilePath = join(
  here, '..', '..', '..', 'packages', 'schema', 'resume.schema.json',
);
const tscBin = join(here, '..', 'node_modules', '.bin', 'tsc');

type SectionTypeLiteral = Section['sectionType'];

// The literal set both dispatch() and the compile-time fixture below are
// written against. Kept as a plain array (not derived from the Section
// type, which TypeScript can't turn into a runtime value) so it can be
// compared against the independently-read schema file at runtime, and
// used to generate the compile-time exhaustiveness fixture.
const KNOWN_SECTION_TYPES: readonly SectionTypeLiteral[] = [
  'profile',
  'work',
  'education',
  'skill',
  'language',
  'certificate',
  'project',
  'custom',
] as const;

// dispatch is the exhaustive switch a real production consumer of Section
// must keep in sync with the generated union — apps/web has no section
// renderer wired up yet (see design spec §5: components/resume/ is a
// later phase), so this test file is where that discipline is enforced
// for now. Every case matches a literal from the real generated Section
// union, never a web-local copy of the discriminator strings.
function dispatch(section: Section): string {
  switch (section.sectionType) {
    case 'profile':
      return 'profile';
    case 'work':
      return 'work';
    case 'education':
      return 'education';
    case 'skill':
      return 'skill';
    case 'language':
      return 'language';
    case 'certificate':
      return 'certificate';
    case 'project':
      return 'project';
    case 'custom':
      return 'custom';
    default: {
      // Compile-time exhaustiveness (documentation only here — the actual
      // enforcement is the tsc-on-a-fixture test below, since vitest does
      // not type-check): if Section ever grows a variant not handled
      // above, `section` is no longer assignable to `never`.
      const exhaustive: never = section;
      throw new Error(
        `schema-contract: unhandled sectionType ${JSON.stringify(exhaustive)}`,
      );
    }
  }
}

function minimalSection(sectionType: SectionTypeLiteral): Section {
  // A minimal, draft-permissive section+entry (design spec §3: only `id`
  // and the section's own sectionType are required). Cast through
  // `as Section` per variant is unnecessary here because the
  // discriminant literal picks the matching union member structurally.
  return {
    sectionType,
    displayName: 'Section',
    iconKey: 'star',
    entries: [{ id: '018f0000-0000-7000-8000-000000000099' }],
  } as Section;
}

interface SchemaSectionOneOfBranch {
  properties?: {
    sectionType?: {
      const?: unknown;
    };
  };
}

function sectionTypesFromSchemaFile(): string[] {
  const raw = readFileSync(schemaFilePath, 'utf8');
  const schema = JSON.parse(raw) as {
    $defs?: { section?: { oneOf?: SchemaSectionOneOfBranch[] } };
  };
  const oneOf = schema.$defs?.section?.oneOf;
  if (!Array.isArray(oneOf) || oneOf.length === 0) {
    throw new Error(
      `schema-contract: ${schemaFilePath} `
      + '$defs.section.oneOf is missing or empty',
    );
  }
  return oneOf.map((branch, index) => {
    const sectionType = branch.properties?.sectionType?.const;
    if (typeof sectionType !== 'string') {
      throw new Error(
        `schema-contract: ${schemaFilePath} $defs.section.oneOf[${index}] `
        + 'has no properties.sectionType.const',
      );
    }
    return sectionType;
  });
}

// exhaustivenessFixtureSource generates a standalone .ts source: a
// dispatch switch with one case per `sectionTypes` entry and a
// `never`-typed default branch, importing the real `Section` type. Kept
// as a generator (not a static file) so the fixture's case list always
// matches KNOWN_SECTION_TYPES exactly, with no risk of the two drifting.
function exhaustivenessFixtureSource(
  sectionTypes: readonly string[],
): string {
  const cases = sectionTypes
    .map((t) => {
      const literal = JSON.stringify(t);
      return `    case ${literal}:\n      return ${literal};`;
    })
    .join('\n');
  return [
    'import type { Section } from \'@aboutme/schema\';',
    '',
    'function dispatch(section: Section): string {',
    '  switch (section.sectionType) {',
    cases,
    '    default: {',
    '      const exhaustive: never = section;',
    '      throw new Error(String(exhaustive));',
    '    }',
    '  }',
    '}',
    'void dispatch;',
    '',
  ].join('\n');
}

describe('schema contract: apps/web recognizes every sectionType', () => {
  it('reads at least one sectionType from resume.schema.json', () => {
    expect(sectionTypesFromSchemaFile().length).toBeGreaterThan(0);
  });

  it('KNOWN_SECTION_TYPES matches resume.schema.json (both directions)', () => {
    const schemaTypes = [...sectionTypesFromSchemaFile()].sort();
    const knownTypes = [...KNOWN_SECTION_TYPES].sort();
    expect(
      schemaTypes,
      'resume.schema.json declares a sectionType set that no longer '
      + 'matches this file\'s KNOWN_SECTION_TYPES / dispatch() — a '
      + 'sectionType was added to (or removed from) the schema without '
      + 'updating this file',
    ).toEqual(knownTypes);
  });

  it.each(KNOWN_SECTION_TYPES)(
    'dispatch() handles a minimal %s section at runtime',
    (sectionType) => {
      expect(dispatch(minimalSection(sectionType))).toBe(sectionType);
    },
  );

  it('every schema sectionType is one dispatch() actually handles', () => {
    for (const sectionType of sectionTypesFromSchemaFile()) {
      expect(
        () => dispatch(minimalSection(sectionType as SectionTypeLiteral)),
        `apps/web does not recognize sectionType ${JSON.stringify(sectionType)}`
        + ' declared by resume.schema.json',
      ).not.toThrow();
    }
  });

  it('KNOWN_SECTION_TYPES exhausts Section at compile time (tsc)', () => {
    const fixturePath = join(here, 'schema-contract-exhaustiveness.tmp.ts');
    const fixtureSource = exhaustivenessFixtureSource(KNOWN_SECTION_TYPES);
    writeFileSync(fixturePath, fixtureSource);
    try {
      expect(() =>
        execFileSync(
          tscBin,
          [
            '--ignoreConfig',
            '--noEmit',
            '--strict',
            '--module', 'esnext',
            '--target', 'esnext',
            '--moduleResolution', 'bundler',
            fixturePath,
          ],
          { stdio: 'pipe' },
        ),
      ).not.toThrow();
    } finally {
      rmSync(fixturePath, { force: true });
    }
  });
});

File: apps/server/internal/api/schema_contract_test.go
// schema_contract_test.go checks the server side of cross-language schema
// conformance. It imports github.com/dannyota/aboutme/packages/schema/gen/go
// as a compile-time dependency, wired through the root go.work and the server's
// go.mod, and independently re-derives the sectionType
// list from resume.schema.json (deliberately not importing anything from
// packages/schema/scripts/generate.mjs or trusting conformance.test.ts's
// derivation — the point is to check the pipeline against the schema file
// itself, the same discipline conformance.test.ts already applies).
//
// It checks two failure modes:
//   - TestKnownSectionTypeConstantsCompile references every
//     schema.SectionType constant apps/server depends on by name. If
//     packages/schema ever renames or removes one, apps/server fails to
//     *compile* (go build ./... and go vet ./... both break), not just
//     fails a test.
//   - TestServerRecognizesEverySchemaSectionType reads resume.schema.json at
//     runtime and fails if the schema declares a sectionType this file's
//     dispatch() doesn't have a case for — the scenario a 9th, newly added
//     sectionType produces: a new schema.SectionType constant exists (it's
//     regenerated from the schema by packages/schema/scripts/generate.mjs),
//     but nothing yet forces dispatch() to add a case for it, so this test
//     fails loudly instead of the server silently ignoring the new type.
package api_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"testing"

	schema "github.com/dannyota/aboutme/packages/schema/gen/go"
)

// schemaFilePath resolves resume.schema.json relative to this source file's
// package directory (apps/server/internal/api), four levels up to the repo
// root (internal/api -> internal -> apps/server -> apps -> repo root) then
// into packages/schema. Valid both for local `go test` (cwd = this
// package's directory) and inside the container build, since
// deploy/server.Dockerfile's build context is the repo root and preserves
// this same relative layout (see deploy/server.Dockerfile, deploy/compose.yml).
func schemaFilePath() string {
	return filepath.Join("..", "..", "..", "..", "packages", "schema", "resume.schema.json")
}

// resumeSchemaSectionOneOf mirrors just enough of resume.schema.json's shape
// to read $defs.section.oneOf[].properties.sectionType.const — the same
// field packages/schema/scripts/generate.mjs's deriveSectionVariants and
// packages/schema/test/conformance.test.ts's sectionTypesFromSchema both
// read, kept as an independent third reader here rather than a shared
// import so this test doesn't trust either of those derivations.
type resumeSchemaSectionOneOf struct {
	Defs struct {
		Section struct {
			OneOf []struct {
				Properties struct {
					SectionType struct {
						Const string `json:"const"`
					} `json:"sectionType"`
				} `json:"properties"`
			} `json:"oneOf"`
		} `json:"section"`
	} `json:"$defs"`
}

// sectionTypesFromSchemaFile reads resume.schema.json off disk and returns
// every sectionType it currently declares, sorted for deterministic
// comparison.
func sectionTypesFromSchemaFile(t *testing.T) []string {
	t.Helper()

	path := schemaFilePath()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("reading %s: %v", path, err)
	}

	var doc resumeSchemaSectionOneOf
	if err := json.Unmarshal(data, &doc); err != nil {
		t.Fatalf("parsing %s: %v", path, err)
	}
	if len(doc.Defs.Section.OneOf) == 0 {
		t.Fatalf("%s: $defs.section.oneOf is missing or empty", path)
	}

	types := make([]string, 0, len(doc.Defs.Section.OneOf))
	for i, branch := range doc.Defs.Section.OneOf {
		if branch.Properties.SectionType.Const == "" {
			t.Fatalf("%s: $defs.section.oneOf[%d] has no properties.sectionType.const", path, i)
		}
		types = append(types, branch.Properties.SectionType.Const)
	}
	sort.Strings(types)
	return types
}

// dispatch is the exhaustive switch a production consumer of schema.Section
// must keep in sync with schema.SectionType. Every case references a generated
// constant by name, never a
// server-local copy of the discriminator string. This keeps internal/resume
// from introducing a parallel discriminator model.
func dispatch(t schema.SectionType) (recognized bool) {
	switch t {
	case schema.Profile,
		schema.Work,
		schema.Education,
		schema.Skill,
		schema.Language,
		schema.Certificate,
		schema.Project,
		schema.SectionTypeCustom:
		return true
	default:
		return false
	}
}

// TestKnownSectionTypeConstantsCompile is the compile-time half of the
// tripwire. Every schema.SectionType constant apps/server currently relies
// on is named here explicitly; if packages/schema ever renames or removes
// one, this file fails to compile, breaking `go build ./...` and
// `go vet ./...`, not merely a test run.
func TestKnownSectionTypeConstantsCompile(t *testing.T) {
	t.Parallel()

	known := []schema.SectionType{
		schema.Profile,
		schema.Work,
		schema.Education,
		schema.Skill,
		schema.Language,
		schema.Certificate,
		schema.Project,
		schema.SectionTypeCustom,
	}
	for _, sectionType := range known {
		if !dispatch(sectionType) {
			t.Fatalf("dispatch() does not recognize its own known constant %q — dispatch() and this list have drifted", sectionType)
		}
	}
}

// TestServerRecognizesEverySchemaSectionType is the runtime half: it reads
// resume.schema.json directly and fails if the schema declares a
// sectionType dispatch() doesn't handle, or if dispatch()'s known set
// contains a sectionType the schema no longer declares. Either direction
// means apps/server and packages/schema have drifted — exactly the failure
// mode a ninth sectionType added to the schema without server-side wiring
// produces.
func TestServerRecognizesEverySchemaSectionType(t *testing.T) {
	t.Parallel()

	schemaTypes := sectionTypesFromSchemaFile(t)

	known := []schema.SectionType{
		schema.Profile,
		schema.Work,
		schema.Education,
		schema.Skill,
		schema.Language,
		schema.Certificate,
		schema.Project,
		schema.SectionTypeCustom,
	}
	knownStrings := make([]string, len(known))
	for i, k := range known {
		knownStrings[i] = string(k)
	}
	sort.Strings(knownStrings)

	if len(schemaTypes) != len(knownStrings) {
		t.Fatalf("resume.schema.json declares %d sectionType(s) %v, but apps/server's dispatch recognizes %d %v — "+
			"a sectionType was added to (or removed from) the schema without updating dispatch() in this file",
			len(schemaTypes), schemaTypes, len(knownStrings), knownStrings)
	}
	for i := range schemaTypes {
		if schemaTypes[i] != knownStrings[i] {
			t.Fatalf("resume.schema.json's sectionType set %v does not match apps/server's dispatch set %v",
				schemaTypes, knownStrings)
		}
	}

	for _, want := range schemaTypes {
		want := want
		t.Run(want, func(t *testing.T) {
			t.Parallel()
			if !dispatch(schema.SectionType(want)) {
				t.Fatalf("apps/server does not recognize sectionType %q declared by resume.schema.json — "+
					"dispatch() in this file needs a case added for it (design spec §3, Codegen fidelity)", want)
			}
		})
	}
}
